### PR TITLE
Cycle tag

### PR DESCRIPTION
**Pull request by Spadavecchia**

### CODE REVIEW

File: lib/liquid/combinators/general.ex [hunk 1]
```diff
@@ -3,6 +3,14 @@ defmodule Liquid.Combinators.General do
...
   """
   import NimbleParsec
+  alias Liquid.Combinators.LexicalToken
+
+  @type comparison_operators :: :== | :!= | :> | :< | :>= | :<= | :contains
+  @type condition :: [
+          condition: {LexicalToken.value(), comparison_operators(), LexicalToken.value()}
+        ]
+  @type liquid_variable :: [liquid_variable: LexicalToken.variable_value(), filters: filter()]
+  @type filter :: [filter: String.t(), params: [values: LexicalToken.value()]]
```
Comment (Eptis) [inline lib/liquid/combinators/general.ex]: Whats the added value for adding a `values` keyword/`{:values, []}` tuples if the entire list contains only LexicalToken values?

Comment (Eptis) [inline lib/liquid/combinators/general.ex]: If `filter` is singular it shouldn't be a list.

File: lib/liquid/combinators/general.ex [hunk 2]
```diff
@@ -3,6 +3,14 @@ defmodule Liquid.Combinators.General do
   General purpose combinators used by almost every other combinator
   """
   import NimbleParsec
+  alias Liquid.Combinators.LexicalToken
+
+  @type comparison_operators :: :== | :!= | :> | :< | :>= | :<= | :contains
+  @type condition :: [
+          condition: {LexicalToken.value(), comparison_operators(), LexicalToken.value()}
+        ]
+  @type liquid_variable :: [liquid_variable: LexicalToken.variable_value(), filters: filter()]
```
Comment (Eptis) [inline lib/liquid/combinators/general.ex]: Shouldn't `filters` be a list?

File: lib/liquid.ex [hunk 3]
```diff
@@ -1,6 +1,26 @@
 defmodule Liquid do
   use Application
 
+  @type t :: [
```
Comment (Eptis) [inline lib/liquid.ex]: These are all tag types, why should they be `Liquid` types?

File: lib/liquid/combinators/tags/case.ex [hunk 4]
```diff
@@ -23,6 +23,13 @@ defmodule Liquid.Combinators.Tags.Case do
   alias Liquid.Combinators.{Tag, General}
   alias Liquid.Combinators.Tags.Generic
 
+  @type t :: [
```
Comment (Eptis) [inline lib/liquid/combinators/tags/case.ex]: Shouldn't a Case type be just a tagged tuple instead of a list of tagged tuples?

File: lib/liquid/combinators/tags/cycle.ex [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    one
+    two
+    three
+    one
+  ```
+  """
+  import NimbleParsec
+  alias Liquid.Combinators.{Tag, General}
+
+  @type t :: [cycle: [group: String.t(), values: [LexicalToken.value()]]]
```
Comment (Eptis) [inline lib/liquid/combinators/tags/cycle.ex]: Shouldn't a Cycle type be just a tagged tuple instead of a list of tagged tuples?

File: lib/liquid/combinators/tags/decrement.ex [hunk 6]
```diff
@@ -23,6 +23,8 @@ defmodule Liquid.Combinators.Tags.Decrement do
   import NimbleParsec
   alias Liquid.Combinators.Tag
 
+  @type t :: [decrement: [variable_name: String.t()]]
```
Comment (Eptis) [inline lib/liquid/combinators/tags/decrement.ex]: Shouldn't a Decrement type be just a tagged tuple instead of a list of tagged tuples?

File: lib/liquid/combinators/tags/for.ex [hunk 7]
```diff
@@ -51,6 +51,22 @@ defmodule Liquid.Combinators.Tags.For do
   alias Liquid.Combinators.{General, Tag}
   alias Liquid.Combinators.Tags.Generic
 
+  @type t :: [
```
Comment (Eptis) [inline lib/liquid/combinators/tags/for.ex]: Shouldn't a For type be just a tagged tuple instead of a list of tagged tuples?

File: lib/liquid/combinators/tags/generic.ex [hunk 8]
```diff
@@ -7,5 +7,9 @@ defmodule Liquid.Combinators.Tags.Generic do
   """
   alias Liquid.Combinators.Tag
 
+  @type else_tag :: [
```
Comment (Eptis) [inline lib/liquid/combinators/tags/generic.ex]: Shouldn't an Else type be just a tagged tuple instead of a list of tagged tuples?

File: lib/liquid/combinators/tags/increment.ex [hunk 9]
```diff
@@ -19,6 +19,8 @@ defmodule Liquid.Combinators.Tags.Increment do
   import NimbleParsec
   alias Liquid.Combinators.Tag
 
+  @type t :: [increment: [variable_name: String.t()]]
```
Comment (Eptis) [inline lib/liquid/combinators/tags/increment.ex]: Shouldn't an Increment type be just a tagged tuple instead of a list of tagged tuples?

File: lib/liquid/nimble_parser.ex [hunk 10]
```diff
@@ -68,6 +69,11 @@ defmodule Liquid.NimbleParser do
   defparsec(:decrement, Decrement.tag())
   defparsec(:increment, Increment.tag())
 
+  defparsec(:cycle_group, Cycle.cycle_group())
```
Comment (Eptis) [inline lib/liquid/nimble_parser.ex]: Does the parser really need all the parsecs for group/body/values? Because I think the cycle parsec is what composes these.

File: lib/liquid/combinators/tags/case.ex [hunk 11]
```diff
@@ -23,6 +23,13 @@ defmodule Liquid.Combinators.Tags.Case do
...
+          case: [
+            variable: LexicalToken.value(),
+            clauses: [String.t() | [when: [conditions: [body: Liquid.t()]]]]
```
Comment (Eptis) [inline lib/liquid/combinators/tags/case.ex]: Based on the tests, I believe this type is incorrect. The conditions aren't a list of tagged :body tuples.
I also think this could be semantically improved to:
```
clauses: [String.t() | [when: condition, then: Liquid.t()]]
```
Not that condition is singular.

File: lib/liquid/combinators/general.ex (hunk 1 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/general.ex]: Every part in the AST have a tag. We can optimize the representation without tags but then we need the context and it is harder. This also makes it consistent with the rest of the AST. We do not want to have separate lists without tags. However, there is an error, it is _value:_, not _values:_

Corrected.

Comment (Spadavecchia) [reply lib/liquid/combinators/general.ex]: You're right, it should be a list. Corrected

File: lib/liquid/combinators/general.ex (hunk 2 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/general.ex]: You're right, `fitlers` must be a list of type `filter`

Corrected.

File: lib/liquid.ex (hunk 3 above)
Comment (Spadavecchia) [reply lib/liquid.ex]: Because the parser is recursive. Almost every _block_ tag allows strings, variables or tags inside which are represented by `Liquid.t()`

File: lib/liquid/combinators/tags/case.ex (hunk 4 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/case.ex]: Every tag is a keyword list (which is a list of tagged tuples)

File: lib/liquid/combinators/tags/cycle.ex (hunk 5 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/cycle.ex]: Every tag is a keyword list (which is a list of tagged tuples)

File: lib/liquid/combinators/tags/decrement.ex (hunk 6 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/decrement.ex]: Every tag is a keyword list (which is a list of tagged tuples)

File: lib/liquid/combinators/tags/for.ex (hunk 7 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/for.ex]: Every tag is a keyword list (which is a list of tagged tuples)

File: lib/liquid/combinators/tags/generic.ex (hunk 8 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/generic.ex]: Every tag is a keyword list (which is a list of tagged tuples)

File: lib/liquid/combinators/tags/increment.ex (hunk 9 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/increment.ex]: Every tag is a keyword list (which is a list of tagged tuples)

File: lib/liquid/nimble_parser.ex (hunk 10 above)
Comment (Spadavecchia) [reply lib/liquid/nimble_parser.ex]: You're right, `group` and `body` can be called like private functions inside `cycle.ex` (corrected) but `values` is recursive and must be defined with `defparsec`

File: lib/liquid/combinators/tags/case.ex (hunk 11 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/case.ex]: You're right, condition must be defined as a new type. I'm going to discuss with Samuel how we can define the `condition` type. 

My first idea when changing `body` for `then` was "I like it" but now I think that I prefer the consistency of `body` tag. Every tag which returns a `Liquid.t()` has a `body` (maybe `block` could be more appropriate)

File: lib/liquid.ex (hunk 3 above)
Comment (Eptis) [reply lib/liquid.ex]: Sorry let me rephrase the question:
Why should those form a public Liquid type and where would it get referenced outside of the parser? I don't see a need for it here, perhaps it fits better as the return type of Parser.parse?

Comment (Spadavecchia) [reply lib/liquid.ex]: `Parser.parse` returns `[String.t() | some_tag() | liquid_varible]` which is the type described in `Liquid` module, however, we need a place to _declare_ the type and I think that the perfect place is inside `Liquid` module because this is the abstraction of what can be processed by liquid. Maybe other place can be the `Template` module.
